app/services/articles/detect_animated_images.rb

```diff
@@ -0,0 +1,36 @@
+module Articles
+  module DetectAnimatedImages
+    IMAGES_IN_LIQUID_TAGS_SELECTORS = [
+      ".liquid-comment img", # CommentTag
+      ".ltag-github-readme-tag img", # GithubReadmeTag
+      ".ltag__link__servicename img", # MediumTag
+      ".ltag__link__taglist img", # LinkTag
+      ".ltag__reddit--container img", # RedditTag
+      ".ltag__stackexchange--container img", # StackexchangeTag
+      ".ltag__twitter-tweet img", # TweetTag
+      ".ltag__user img", # UserTag and OrganizationTag
+      ".ltag__user-subscription-tag img", # UserSubscriptionTag
+      ".ltag_github-liquid-tag img", # GitHubIssueTag
+      ".podcastliquidtag img", # PodcastTag
+    ].join(", ").freeze
+
+    def self.call(article)
+      parsed_html = Nokogiri::HTML.fragment(article.processed_html)
+
+      # we ignore images contained in liquid tags as they are not animated
```

Comment:
Great question! "Always be true" is never true for almost anything. In this context I think we can be reasonably confident or at least move forward with the status quo and then improve in the future, mainly for two reasons.

**Liquid tags we control**

Out of all the liquid tags we control (those rendering inner content or those implemented as HTML) we only display static images or profile images:

<details>
<summary>Grep of liquid tags templates</summary>

```shell
app/views/articles/_liquid.html.erb
6:        <img src="<%= asset_path("video-camera.svg") %>" alt="video camera">
12:      <img src='<%= Images::Profile.call(article.user.profile_image_url, length: 150) %>' alt='<%= "#{article.user.username} image" %>'>

app/views/organizations/_liquid.html.erb
11:      <img src="<%= Images::Profile.call(organization.profile_image_url, length: 150) %>" alt="<%= "#{organization.slug} image" %>" />

app/views/comments/_liquid.html.erb
5:        <img class="profile-pic" src="<%= Images::Profile.call(comment.user.profile_image_url, length: 50) %>" alt="<%= comment.user.username %> profile image" />

app/views/podcast_episodes/_liquid.html.erb
1:<% image_url = Images::Optimizer.call(podcast.pattern_image_url || "https://i.imgur.com/fKYKgo4.png", crop: nil, fetch_format: "jpg") %>
16:    <img class="button play-butt" id="play-butt-<%= episode.slug %>" src="<%= asset_path("playbutt.png") %>" alt="play">
17:    <img class="button pause-butt" id="pause-butt-<%= episode.slug %>" src="<%= asset_path("pausebutt.png") %>" alt="pause">

app/views/users/_liquid.html.erb
12:        <img src="<%= Images::Profile.call(user.profile_image_url, length: 150) %>" alt="<%= "#{user.username} image" %>" />
17:      <img src="<%= Images::Profile.call(user.profile_image_url, length: 150) %>" alt="<%= "#{user.username} image" %>" />

_reddit.html.erb
6:          <img src="<%= asset_path("reddit-icon.svg") %>" alt="Reddit Logo" />
21:        <img src="<%= thumbnail %>" alt="<%= title %>">

_stackexchange.html.erb
7:          <img src="<%= ActionController::Base.helpers.asset_path(logo_to_use) %>" alt="">
21:        <img src="<%= ActionController::Base.helpers.asset_path("stackexchange-arrow-up.svg") %>" alt="">
25:        <img src="<%= ActionController::Base.helpers.asset_path("stackexchange-arrow-down.svg") %>" alt="">

_github_readme.html.erb
4:      <img src="<%= ActionController::Base.helpers.asset_path("github-logo.svg") %>" alt="GitHub logo">

_medium.html.erb
4:      <img src='<%= response[:author_image] %>' alt='<%= response[:author] %>' />
12:        <img src="<%= asset_path("medium_icon.svg") %>" alt="Medium Logo" aria-label="Medium Logo" />

_tweet.html.erb
6:          <img src="<%= tweet.extended_entities_serialized[:media].first[:media_url_https] %>" alt="unknown tweet media content">
7:          <img src="/assets/play-butt.svg" class="ltag__twitter-tweet__play-butt" alt="Play butt">
18:        <img src="<%= tweet.extended_entities_serialized[:media].first[:media_url_https] %>" alt="unknown tweet media content">
23:      <img src="<%= tweet.extended_entities_serialized[:media].first[:media_url_https] %>" alt="unknown tweet media content">
29:      <img class="ltag__twitter-tweet__profile-image"
39:        <img src="<%= twitter_logo %>" alt="twitter logo">
63:        <img src="<%= ActionController::Base.helpers.asset_path("twitter-reply-action.svg") %>" alt="Twitter reply action">
66:        <img src="<%= ActionController::Base.helpers.asset_path("twitter-retweet-action.svg") %>" alt="Twitter retweet action">
69:        <img src="<%= ActionController::Base.helpers.asset_path("twitter-like-action.svg") %>" alt="Twitter like action">

_wikipedia.html.erb
3:    <img src="<%= ActionController::Base.helpers.asset_path("wikipedia-logo.svg") %>" class="ltag__wikipedia--logo" alt="Wikipedia Logo" aria-label="Wikipedia Logo">

_github_issue.html.erb
4:      <img class="github-logo" alt="GitHub logo" src="<%= ActionController::Base.helpers.asset_path("github-logo.svg") %>">
14:        <img class="github-liquid-tag-img" src="<%= user_avatar_url %>" alt="<%= username %> avatar">

_user_subscription.html.erb
9:          <img class="crayons-avatar__image ltag__user-subscription-tag__author-profile-image" src=<%= author_profile_image %> <%= "#{author_username} profile image" %>>
13:          <img class="crayons-avatar__image ltag__user-subscription-tag__subscriber-profile-image">
```
</details>

In this scenario there two cases to be mindful about: 

- profile images which technically could include GIFs (there [only a few on DEV](https://github.com/forem/rfcs/issues/125#issuecomment-783586182) but not zero)
- moving images on Twitter, this is not likely as Twitter converts images into mp4 videos

The technique explored in this Spike could be also applied to profile images and cover images, but we can get there one step at a time

**Liquid tags we don't control**

The other scenario is `<iframe>`s: most liquid tags are iframes. I tried to render a few of them and it's still unlikely people will encounter gifs BUT there's a big unknown as I could technically embed a Codepen with a gif. But being an iframe, it's still something we can't do anything about on the server.

So this solution doesn't cover ALL scenarios but I think it's good enough for most of them and it can be improved upon

Reply:
> profile images which technically could include GIFs (there only a few on DEV but not zero)

Maybe worth mentioning here that we have [another stage 0 RFC](https://github.com/forem/rfcs/issues/125) to prevent users from using animated images as their profile avatar. If we follow through on that then we won't have the issue of animated profile images inside liquid tags in future 🙂